### PR TITLE
Constant folding for dynamic shape node

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129686
* #130367

Extend constant folding for dynamic shape node, only support pointwise op and some restricted ops

We support dynamic shapes by limiting constant folding of ops that are guaranteed to have uniform values (full, pointwise ops, and views) and running these operators with tensors of shape 1. This also eliminates the possibility of memory overhead of constant folding.

Taken over from https://github.com/pytorch/pytorch/pull/128937 

joint work with @imzhuhl  

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/constant_folding.py [hunk 1]
```diff
@@ -60,6 +60,13 @@ def __init__(
         # is the output
         self.user_to_last_uses = self.node_to_last_non_output_use()
 
+    def _support_dynamic_shape(self):
+        # ConstantFolder not support dynamic shape now
```
Comment (Chillee) [inline torch/_inductor/constant_folding.py:64]: nit: grammar lol

File: torch/_inductor/constant_folding.py [hunk 2]
```diff
@@ -194,10 +203,13 @@ def add_node_replacement(self, node: torch.fx.Node, tensor: torch.Tensor) -> Non
         self.node_replacements[node] = tensor
 
     def run(self):
-        env = {}
+        env: Dict[torch.fx.Node, Any] = {}
+        self.insert_placerholder_values(env)
+        return super().run(initial_env=env)
+
+    def insert_placerholder_values(self, env: Dict[torch.fx.Node, Any]) -> None:
```
Comment (Chillee) [inline torch/_inductor/constant_folding.py:210]: nit: spelling

File: torch/_inductor/fx_passes/joint_graph.py [hunk 3]
```diff
@@ -201,21 +203,107 @@ def __init__(self, gm, skip_constructors=False):
...
+            aten.expand,
+            aten.as_strided,
+            aten.permute,
+        ]
+
+    def _support_dynamic_shape(self):
+        return True
+
     def insertable_tensor_check(self, t: torch.Tensor) -> bool:
         # TODO - we could also Tensors which get replaced with arange here
```
Comment (Chillee) [inline torch/_inductor/fx_passes/joint_graph.py:226]: comment outdated?

File: torch/_inductor/fx_passes/joint_graph.py [hunk 4]
```diff
@@ -201,21 +203,107 @@ def __init__(self, gm, skip_constructors=False):
         # see: [constant folding refining of symints]
         self.node_replacements_shapes: Dict[torch.fx.Node, List[int]] = {}
 
+        self.symint_nodes = _SymHashingDict()
+
+        # reference from torch/_funtorch/partitioners.py:get_default_op_list
```
Comment (Chillee) [inline torch/_inductor/fx_passes/joint_graph.py:208]: We need tags for these lol

**Review by Chillee — APPROVED** [review]
some nits

File: torch/_inductor/constant_folding.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/constant_folding.py:210]: ?

File: torch/_inductor/fx_passes/joint_graph.py (hunk 4 above)
Comment (eellison) [reply torch/_inductor/fx_passes/joint_graph.py:208]: Yea https://github.com/pytorch/pytorch/issues/129020..

**Comment by atalman** [comment]
This one is also failing: https://github.com/pytorch/ao/blob/main/test/prototype/mx_formats/test_mx_tensor.py

**Pull request closed without merging**